securedrop/tests/test_source.py

```diff
@@ -74,462 +72,489 @@ def _find_codename(self, html):
+                logger.assert_called_once()
+                assert ("Attempt to create a source with duplicate codename"
+                        in logger.call_args[0][0])
+                assert 'codename' not in session
 
-    def test_lookup(self):
+    def test_lookup(self, source_app):
         """Test various elements on the /lookup page."""
-        with self.client as client:
-            codename = new_codename(client, session)
-            resp = client.post('login', data=dict(codename=codename),
-                               follow_redirects=True)
+        with source_app.test_client() as app:
+            codename = new_codename(app, session)
+            resp = app.post('/login', data=dict(codename=codename),
+                            follow_redirects=True)
             # redirects to /lookup
-            self.assertIn("public key", resp.data)
+            text = resp.data.decode('utf-8')
+            assert "public key" in text
             # download the public key
-            resp = client.get('journalist-key')
-            self.assertIn("BEGIN PGP PUBLIC KEY BLOCK", resp.data)
-
-    def test_login_and_logout(self):
-        resp = self.client.get('/login')
-        self.assertEqual(resp.status_code, 200)
-        self.assertIn("Enter Codename", resp.data)
-
-        with self.client as client:
-            codename = new_codename(client, session)
-            resp = client.post('/login', data=dict(codename=codename),
-                               follow_redirects=True)
-            self.assertEqual(resp.status_code, 200)
-            self.assertIn("Submit Materials", resp.data)
-            self.assertTrue(session['logged_in'])
-            resp = client.get('/logout', follow_redirects=True)
-
-        with self.client as c:
-            resp = c.post('/login', data=dict(codename='invalid'),
-                          follow_redirects=True)
-            self.assertEqual(resp.status_code, 200)
-            self.assertIn('Sorry, that is not a recognized codename.',
-                          resp.data)
-            self.assertNotIn('logged_in', session)
-
-        with self.client as c:
-            resp = c.post('/login', data=dict(codename=codename),
-                          follow_redirects=True)
-            self.assertEqual(resp.status_code, 200)
-            self.assertTrue(session['logged_in'])
-            resp = c.get('/logout', follow_redirects=True)
-
-            # sessions always have 'expires', so pop it for the next check
-            session.pop('expires', None)
-
-            self.assertNotIn('logged_in', session)
-            self.assertNotIn('codename', session)
-
-            self.assertIn('Thank you for exiting your session!', resp.data)
+            resp = app.get('/journalist-key')
+            text = resp.data.decode('utf-8')
+            assert "BEGIN PGP PUBLIC KEY BLOCK" in text
+
+    def test_login_and_logout(self, source_app):
+        with source_app.test_client() as app:
```

Comment:
Why is the grounds for that nit? I think it might do some silly things to sessions and whatnot (but could be wrong).

Reply:
Since the code in each block is a separate test case, having them use a different test client makes sense (really they should be in different tests entirely - but in the spirit of not expanding the scope of this PR beyond strictly rewriting the tests to use pytest fixtures, @dachary's suggestion seems reasonable)